cbits/src/mimalloc.rs

```diff
@@ -0,0 +1,289 @@
+    fn mi_realloc_aligned(p: *mut c_void, newsize: usize, alignment: usize) -> *mut c_void;
+
+    /// Free any previously allocated memory.
+    fn mi_free(p: *mut c_void);
+
+    /// Merge all thread-local statistics with the main statistics buffer, and
+    /// reset.
+    fn mi_stats_merge();
+
+    /// Reset the internal statistics counters.
+    fn mi_stats_reset();
+
+    /// Print global mimalloc statistics to a given output. The output is
+    /// currently fixed to `stderr`; both the first and second arguments must be
+    /// the NULL pointer.
+    fn mi_stats_print_out(out: unsafe extern "C" fn(*const c_char, *mut c_void), ctx: *mut c_void);
+
+    /// Register a "deferred free" function which will be called
+    /// deterministically over the course of the program.
+    fn mi_register_deferred_free(
+        f: unsafe extern "C" fn(bool, c_ulonglong, *mut c_void),
+        ctx: *mut c_void,
+    );
+}
+
+/// Provide global mimalloc heap statistics to a user-provided callback.
+///
+/// The provided callback function will be called multiple times in sequence,
+/// each time with a single argument, which is a single null-terminated line,
+/// represented as a [CStr]. Collectively, these lines will represent a summary
+/// of the global heap statistics for the entire program, meant to be written to
+/// the terminal or a log file.
+///
+/// Therefore, the simplest way to use this function is to simply provide a
+/// closure that prints the given log messages to `stderr` immediately:
+///
+/// ```rust,ignore
+/// eprintln!("========================================");
+/// eprintln!("mimalloc memory allocation statistics:\n");
+/// jj_cbits::mimalloc::stats_print(&|l| {
+///   eprint!("{}", l.to_string_lossy());
+/// });
+/// ```
+///
+/// Note that this merges all thread-local statistics into the main statistics
+/// summary before printing to `stderr`, so while it will give a global summary
+/// of the heap, it may cause some performance overhead while thread-local
+/// buffers are being flushed and merged.
+pub fn stats_print<F: Fn(&CStr)>(f: &'static F) {
+    unsafe {
+        unsafe extern "C" fn wrapper<F: Fn(&CStr)>(value: *const c_char, ctx: *mut c_void) {
+            (*(ctx as *const F))(CStr::from_ptr(value));
+        }
+        mi_stats_merge();
+        mi_stats_print_out(wrapper::<F>, f as *const F as *mut c_void)
+    }
+}
+
+/// Reset heap statistics counters and histograms.
+///
+/// Primarily useful to clear out any existing statistics, so that a subsequent
+/// call to `mimalloc_stats_print` will only show statistics since the last
+/// reset.
+///
+/// This should also reset and merge all thread-local statistics, too.
+pub fn stats_reset() {
+    unsafe {
+        mi_stats_reset();
+    }
+}
+
+/// Register a "deferred free" function, which will be called by the memory
+/// allocator after some (deterministic) number of calls to
+/// [`dealloc`](core::alloc::GlobalAlloc::dealloc) in the heap.
```

Comment:
[Upstream docs](https://microsoft.github.io/mimalloc/group__extended.html#ga3460a6ca91af97be4058f523d3cb8ece) indicate that this is called after *allocations*, not deallocations, which makes a lot more sense.